### ISSUE TITLE
Fingerprints are different on load balancing

### ISSUE THREAD

**Issue by ThorkildFriis** [scandihealth/lpr3-docs#291]
It seems that load balancing provides different fingerprints for each certificate. Either you have to accept each fingerprint or to examine the content of the certificate...

**Comment by ThorkildFriis** [comment]
Thanks, On the live server we sometimes have issues with finger prints from the certificate. We have not seen the issue in test and most of the time it works on the live server. Our service does not recognize the certificate and, hence, it can not trust it. I can not know for sure whether it is based on load balancing. But it is my guess. It works most of the time and it always works in the test environment.

**Comment by TueCN** [comment]
Ok, I assume you are talking about the fingerprint of the SSL certificate sent by https://lpradm.sds.dsdn.dk ?
I first assumed you were saying that LPR did not trust your XML signature of the SOAP header.

I don't fully understand what the log you pasted is for. Is it a log of the entries in your trust store? 
If so, it seems there are 2 entries for _.sds.dsdn.dk_. Could that be the issue?

We just tested by sending 1000 requests to https://lpradm.sds.dsdn.dk and all of them returned a SHA1 fingerprint of `32:DA:7F:86:65:AB:78:1C:EF:B6:92:EE:31:34:BF:74:08:DE:6B:83`, which is as expected. 

We have not heard of anyone else having issues verifying our SSL certificate. 
Are you sure the error is not on your side?

**Comment by ThorkildFriis** [comment]
Today we saw the same issue when submitting 4500 and later 54000 records in test. This is the first time we have seen it in test. So it is perhaps reasonable to assume that the problem can not be reproduced after you successfully submitted 1000 posts. But we did not see that issue before submitting more than 600,000 in ten threads on the live server and before in test before that we never submitted that many records and we only did it from one thread. We did today see it with as little as 4500 records from ten threads... And it is not like only one thread is failing. 
I am not a programmer (used to be though) but it does not make sense to me that our algorithm should (the same piece of code) fail when submitting many records from several threads and not fail when submitting a few records from one thread unless there is an issue with locking and race condition. 
We will, of course, check the race condition option, but I really do not believe that this is the issue. The guy who programmed it is too experienced to make a mistake like that, unless there is a bug in Java that is only present if the word LPR3 is in the name of the project.... I mean, it is the same algorithm that we use for all other communication. We will check, of course. But I suggest that you handle this issue as not "cannot reproduce". Perhaps try with 50,000 or even better 600,000 posts from 10 threads. We have not seen the issue with 1000 records in one thread either.

**Comment by jonigkeit** [comment]
We will perform a test tomorrow.

Since you experience the issue regularly will you please provide us with a TCP dump of the offending TLS handshake.

**Comment by jonigkeit** [comment]
We have completed our test of 12 concurrent calls repeating 50,000 times for a total of 600,000 invocations and every single time the fingerprint remains the same
````
SHA256 Fingerprint=BA:93:8F:74:76:6B:F2:31:83:0A:56:AB:6F:B5:AF:83:CF:52:30:90:05:33:DD:D6:28:E6:BB:D0:B2:F2:A9:BE
````

@ThorkildFriis since you experience the issue regularly will you please provide us with a TCP dump of the offending TLS handshake.

**Comment by ThorkildFriis** [comment]
We tried to add more log and send 100.000 messages. After about 10.000 messages we stopped and examined the log. The proxy we use is written in Java v.6 because of the version of the Websphere Application Server we use. It turns out that the Java library method to check certificates is not thread safe. We added a Synchronize() to the method call and the remainder 90.000 messages went though without any issues.
